src/main/kotlin/org/jitsi/utils/queue/QueueStatistics.kt

```diff
@@ -0,0 +1,225 @@
+     */
+    private val totalPacketsDropped = LongAdder()
+
+    /**
+     * The time the first packet was added.
+     */
+    private var firstPacketAdded: Instant? = null
+
+    /**
+     * Statistics about queue lengths
+     */
+    private val queueLengthStats = BucketStats(getQueueLengthBucketSizes(queueSize), "_queue_size_at_remove", "")
+
+    /**
+     * Statistics about the time that packets were waiting in the queue.
+     */
+    private val queueWaitStats = if (TRACK_TIMES) BucketStats(waitBucketSizes, "_queue_wait_time_ms", " ms") else null
+
+    /**
+     * Gets a snapshot of the stats in JSON format.
+     */
+    val stats: OrderedJsonObject
+        get() {
+            val stats = OrderedJsonObject()
+            val now = clock.instant()
+            stats["added"] = totalPacketsAdded.sum()
+            stats["removed"] = totalPacketsRemoved.sum()
+            stats["dropped"] = totalPacketsDropped.sum()
+            val duration = Duration.between(firstPacketAdded, now)
+            val duration_s = duration.toNanos() / 1e9
+            stats["duration_s"] = duration_s
+            val packetsRemoved = totalPacketsRemoved.sum().toDouble()
+            stats["average_remove_rate_pps"] = packetsRemoved / duration_s
+            stats["queue_size_at_remove"] = queueLengthStats.toJson()
+            queueWaitStats?.let { stats["queue_wait_time"] = it.toJson() }
+            return stats
+        }
+
+    /**
+     * Registers the addition of a packet.
+     */
+    fun added() {
+        if (firstPacketAdded == null) {
+            firstPacketAdded = clock.instant()
+        }
+        totalPacketsAdded.increment()
+    }
+
+    /**
+     * Registers the removal of a packet.
+     */
+    fun removed(queueSize: Int, waitTime: Duration?) {
+        totalPacketsRemoved.increment()
+        queueLengthStats.addValue(queueSize.toLong())
+        if (waitTime != null) {
+            queueWaitStats?.addValue(waitTime.toMillis()) /* TODO: measure in nanos? */
+        }
+    }
+
+    /**
+     * Registers that a packet was dropped.
+     */
+    fun dropped() {
+        totalPacketsDropped.increment()
+    }
+
+    companion object {
+        /** Whether specific per-queue statistics should be kept. */
+        @JvmField
+        @field:SuppressFBWarnings("MS_SHOULD_BE_FINAL")
+        var DEBUG = false
+
+        /** Whether queue dwell times should be tracked. */
+        @JvmField
+        @field:SuppressFBWarnings("MS_SHOULD_BE_FINAL")
+        var TRACK_TIMES = false
+
+        private val queueStatsById = ConcurrentHashMap<String, QueueStatistics>()
+
+        internal fun globalStatsFor(queue: PacketQueue<*>, clock: Clock) = queueStatsById.computeIfAbsent(queue.id()) {
+            /* Assume all queues with the same ID have the same capacity and can use the same size buckets. */
+            QueueStatistics(queue.capacity(), clock)
+        }
+
+        fun getStatistics(): OrderedJsonObject {
```

Comment:
This could be just `= OrderedJsonObject().apply { queueStatsById.entries.forEach { put(it.key, it.value.stats) } }`. This feels easier to read, but I know that's subjective.